### PR TITLE
[dhcp6relay] Support relaying Relay-Forward message

### PR DESCRIPTION
**Pull request by kellyyeh**
#### Why I did it
Allow dhcp6relay to relay DHCPv6 Relay messages.

#### How I did it
Encapsulate Relay-Forward message in a new Relay-Forward message. No changes needed on the Relay-Reply from server.

#### How to verify it
Send a Relay Forward message, check if a new Relay Forward message was constructed and sent to the servers.
Tested with https://github.com/Azure/sonic-mgmt/pull/5051

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106

### CODE REVIEW

File: src/dhcp6relay/src/relay.cpp [hunk 1]
```diff
@@ -468,6 +468,47 @@ void relay_client(int sock, const uint8_t *msg, int32_t len, const ip6_hdr *ip_h
...
+ * @param msg            pointer to dhcpv6 message header position
+ * @param len            size of data received
+ * @param ip_hdr         pointer to IPv6 header
+ * @param ether_hdr      pointer to Ethernet header
+ * @param config         pointer to the relay interface config
+ *
+ * @return none
+ */
+void relay_relay_forw(int sock, const uint8_t *msg, int32_t len, const ip6_hdr *ip_hdr, const ether_header *ether_hdr, relay_config *config) {
+    uint8_t buffer[4096];
```
Comment (yxieca) [inline src/dhcp6relay/src/relay.cpp:486]: Make it static for now. Having it on stack has the risk of stack overflow.

File: src/dhcp6relay/src/relay.cpp [hunk 2]
```diff
@@ -468,6 +468,47 @@ void relay_client(int sock, const uint8_t *msg, int32_t len, const ip6_hdr *ip_h
...
+ * @param sock           L3 socket for sending data to servers
+ * @param msg            pointer to dhcpv6 message header position
+ * @param len            size of data received
+ * @param ip_hdr         pointer to IPv6 header
+ * @param ether_hdr      pointer to Ethernet header
+ * @param config         pointer to the relay interface config
+ *
+ * @return none
+ */
+void relay_relay_forw(int sock, const uint8_t *msg, int32_t len, const ip6_hdr *ip_hdr, const ether_header *ether_hdr, relay_config *config) {
```
Comment (yxieca) [inline src/dhcp6relay/src/relay.cpp:485]: change forw to forward or fwd.

File: src/dhcp6relay/src/relay.h [hunk 3]
```diff
@@ -15,7 +15,7 @@
 
 #define RELAY_PORT 547
 #define CLIENT_PORT 546
-#define HOP_LIMIT 32
+#define HOP_LIMIT 8
```
Comment (yxieca) [inline src/dhcp6relay/src/relay.h:18]: Why changing hop limit?

File: src/dhcp6relay/src/relay.cpp (hunk 2 above)
Comment (yxieca) [inline src/dhcp6relay/src/relay.cpp:485]: ether_hdr is not used?

File: src/dhcp6relay/src/relay.h (hunk 3 above)
Comment (kellyyeh) [reply src/dhcp6relay/src/relay.h:18]: Correcting hop limit according to rfc: https://datatracker.ietf.org/doc/html/rfc8415#:~:text=8.%20%20%20Section%207%20(%22DHCP%20Constants%22)%20was%20primarily%20updated%20to%20add%0A%20%20%20%20%20%20%20%20constants%20from%20RFC%204242%20and%20RFC%207083.%20%20Note%20that%20the%20default%0A%20%20%20%20%20%20%20%20HOP_COUNT_LIMIT%20value%20was%20reduced%20from%2032%20to%208.

File: src/dhcp6relay/src/relay.cpp (hunk 2 above)
Comment (kellyyeh) [reply src/dhcp6relay/src/relay.cpp]: Removed

File: src/dhcp6relay/src/relay.cpp (hunk 1 above)
Comment (kellyyeh) [reply src/dhcp6relay/src/relay.cpp]: Thanks for catching that. Updated

File: src/dhcp6relay/src/relay.h (hunk 3 above)
Comment (yxieca) [reply src/dhcp6relay/src/relay.h:18]: Can you add a comment accordingly?

Comment (kellyyeh) [reply src/dhcp6relay/src/relay.h]: Updated

**Review by yxieca — APPROVED** [review]

**Comment by ppikh** [comment]
@kellyyeh - Can we have this fix in 202111 branch?

**Pull request merged**